simplicity-sys/src/types.rs

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: CC0-1.0
+
+#![allow(non_camel_case_types)]
+
+pub type c_int = i32;
+pub type c_uchar = u8;
+pub type c_uint = u32;
+pub type size_t = usize;
+pub use core::ffi::c_void;
+
+#[cfg(test)]
+mod tests {
+    use crate::types;
+    use std::any::TypeId;
+    use std::os::raw;
+
+    #[test]
+    fn verify_types() {
+        assert_eq!(TypeId::of::<types::c_int>(), TypeId::of::<raw::c_int>());
+        assert_eq!(TypeId::of::<types::c_uchar>(), TypeId::of::<raw::c_uchar>());
+        assert_eq!(TypeId::of::<types::c_uint>(), TypeId::of::<raw::c_uint>());
+    }
+}
+
+#[doc(hidden)]
+#[cfg(target_arch = "wasm32")]
+pub fn sanity_checks_for_wasm() {
+    use core::mem::{align_of, size_of};
+    extern "C" {
+        pub static WASM32_INT_SIZE: c_uchar;
+        pub static WASM32_INT_ALIGN: c_uchar;
+
+        pub static WASM32_UNSIGNED_INT_SIZE: c_uchar;
+        pub static WASM32_UNSIGNED_INT_ALIGN: c_uchar;
+
+        pub static WASM32_SIZE_T_SIZE: c_uchar;
+        pub static WASM32_SIZE_T_ALIGN: c_uchar;
+
+        pub static WASM32_UNSIGNED_CHAR_SIZE: c_uchar;
+        pub static WASM32_UNSIGNED_CHAR_ALIGN: c_uchar;
+
+        pub static WASM32_PTR_SIZE: c_uchar;
+        pub static WASM32_PTR_ALIGN: c_uchar;
+    }
+    unsafe {
+        assert_eq!(size_of::<c_int>(), WASM32_INT_SIZE as usize);
```

Comment:
In 04933cfb72dbb1ff31987f543f3611ed50524427:

I think we should actually make these checks unconditional. So rather than calling them `WASM32_` call them `C_` and always compare the size/alignment as computed by the C compiler against the ones implied by the `type size_t = usize` etc.

Reply:
You can also unconditionally include `wasm.c` which I guess ought to be renamed to `ffi_checks.c` or something.

Reply:
The include is in 076fe4a75f21c5fa62e37f5af0df8d2cde925db1. Also, that patch needs to come before the one that adds the unit tests.

Reply:
Wait -- we already have all this stuff. The constants are in `depend/wrapper.c` (we need to add some new ones for int, pointers and size_t) (we really don't need to test `unsigned_char` this is guaranteed by the C standard to have size and alignment 1).

Then the tests are in `simplicity-sys/src/c_jets/mod.rs` `sanity_checks`.

Why add a new wasm.c rather than just extending the existing stuff?

Reply:
I copied from rust-secp-sys, but it seems we already have some of the functionality. I will remove the duplicates tomorrow.

Reply:
I think rust-secp-sys copied its stuff from here :)